oml/retrieval/retrieval_results.py

```diff
@@ -0,0 +1,182 @@
+        self.gt_ids = gt_ids
+
+    @property
+    def n_retrieved_items(self) -> int:
+        return self.retrieved_ids.shape[1]
+
+    @classmethod
+    def compute_from_embeddings(
+        cls,
+        embeddings: FloatTensor,
+        dataset: IQueryGalleryDataset,
+        n_items_to_retrieve: int = 1_000,
+    ) -> "RetrievalResults":
+        """
+        Args:
+            embeddings: The result of inference with the shape of ``[dataset_len, emb_dim]``.
+            dataset: Dataset having query/gallery split.
+            n_items_to_retrieve: Number of the closest gallery items to retrieve.
+                                 It may be clipped by gallery size if needed.
+
+        """
+        assert len(embeddings) == len(dataset), "Embeddings and dataset must have the same size."
+
+        # todo 522: rework taking sequence
+        if hasattr(dataset, "df") and SEQUENCE_COLUMN in dataset.df:
+            dataset.df[SEQUENCE_COLUMN], _ = pd.factorize(dataset.df[SEQUENCE_COLUMN])
+            sequence_ids = LongTensor(dataset.df[SEQUENCE_COLUMN])
+        else:
+            sequence_ids = None
+
+        labels_gt = dataset.get_labels() if isinstance(dataset, IQueryGalleryLabeledDataset) else None
+
+        distances, retrieved_ids, gt_ids = batched_knn(
+            embeddings=embeddings,
+            ids_query=dataset.get_query_ids(),
+            ids_gallery=dataset.get_gallery_ids(),
+            labels_gt=labels_gt,
+            sequence_ids=sequence_ids,
+            top_n=n_items_to_retrieve,
+        )
+
+        return RetrievalResults(distances=distances, retrieved_ids=retrieved_ids, gt_ids=gt_ids)
+
+    def __repr__(self) -> str:
+        txt = (
+            f"You retrieved {self.n_retrieved_items} items.\n"
+            f"Distances to the retrieved items:\n{self.distances}.\n"
+            f"Ids of the retrieved gallery items:\n{self.retrieved_ids}.\n"
+        )
+
+        if self.gt_ids is None:
+            txt += "Ground truths are unknown.\n"
+        else:
+            gt_ids_list = [x.tolist() for x in self.gt_ids]
+            txt += f"Ground truth gallery ids are:\n{gt_ids_list}.\n"
+
+        return txt
+
+    def visualize(
+        self,
+        query_ids: List[int],
+        dataset: IQueryGalleryDataset,
+        n_galleries_to_show: int = 5,
+        verbose: bool = False,
+    ) -> plt.Figure:
+        """
+        Args:
+            query_ids: Query indices within the range of ``(0, n_query - 1)``.
+            dataset: Dataset that provides query-gallery split and supports visualisation.
+            n_galleries_to_show: Number of closest gallery items to show.
+            verbose: Set ``True`` to allow prints.
+
+        """
+        if not isinstance(dataset, (IVisualizableDataset, IQueryGalleryDataset)):
+            raise ValueError(
+                f"Dataset has to support {IVisualizableDataset.__name__} and "
+                f"{IQueryGalleryDataset} interfaces. Got {type(dataset)}."
```

Comment:
Add `.__name__` as for `{IVisualizableDataset.__name__}`?

Reply:
oh right